### PR TITLE
Social distancing

### PR DESCRIPTION
**Pull request by jiayuanz3**
## Summary

Describe the issue, and what you have done to solve it

## Further work

If relevant, add in what needs to be done next on this area (and whether you will do it!)

## Checklist

- [ ] All new functions have docstrings in the correct style
- [ ] I've verified the complete docs build locally without errors
- [ ] I've maintained 100% coverage (please mention any 'no cover' annotations explicitly)
- [ ] I've unit-tested all new methods directly

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have read the **CONTRIBUTING** document.
- [ ] My code follows the code style of this project.
- [ ] I have updated the wiki with new parameters/model functionality

## Closing issues

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: pyEpiabm/pyEpiabm/property/household_foi.py [hunk 1]
```diff
@@ -99,7 +99,19 @@ def household_foi(infector, infectee, time: float):
...
+                             intervention_params[
+                             'social_distancing'][
+                             'distancing_house_enhanced_susc']
+            else:
+                distancing = Parameters.instance().\
+                             intervention_params[
+                             'social_distancing'][
+                             'distancing_house_susc']
+        else:
+            distancing = 1
```
Comment (HenrietteCapel) [inline pyEpiabm/pyEpiabm/property/household_foi.py]: I'm not sure if it is correct to put the scaling in 'foi'. According to CovidSim: "If so, then the susceptibility of all people residing in microcell currently undergoing social distancing is scaled, either by enhanced or non-enhanced effects (see below). Infectiousness functions are unaffected by social distancing." So should scaling be in 'household_susc'?

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        super(TestSocialDistancing, cls).setUpClass()  # Sets up parameters
+        cls.time = 1
+        cls._population = pe.Population()
+        cls._population.add_cells(1)
+        cls._population.cells[0].add_microcells(1)
+        cls.microcell = cls._population.cells[0].microcells[0]
+        cls.microcell.add_people(1)
+        cls.person = cls.microcell.persons[0]
+        cls.person.update_status(InfectionStatus(7))
+
```
Comment (HenrietteCapel) [inline pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py:27]: We define a population a bit different in 'test_household_quarantine.py' and I think that way is more clear. Bit confused here because you are only adding one person with a certain InfectionStatus. Is there a second which is susceptible? 

I think this will be more clear:
```
cls.pop_factory = pe.routine.ToyPopulationFactory()
cls.pop_params = {"population_size": 2, "cell_number": 1,
                          "microcell_number": 1, "household_number": 1}
cls.test_population = cls.pop_factory.make_pop(cls.pop_params)
cls.sympt_person = cls.test_population.cells[0].microcells[
            0].persons[0]
cls.sympt_person.update_status(InfectionStatus.InfectMild)
cls.susc_person = cls.test_population.cells[0].microcells[0].persons[1]
cls.susc_person.update_status(InfectionStatus.Susceptible)
```

File: python_examples/intervention_example/social_distancing_simulation_flow.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                              'distancing_spatial_susc': [0.1, 0.3, 0.9]}
+for to_modify_parameter, parameter_values in to_modify_parameter_values.\
+        items():
+    for parameter_value in parameter_values:
+        name_output_file = 'output_{}_{}.csv'.format(
+                parameter_value, to_modify_parameter)
+
+        if to_modify_parameter == 'distancing_enhanced_prob':
+            pe.Parameters.instance().intervention_params['social_distancing'][
+                to_modify_parameter] = [parameter_value] * 17
```
Comment (HenrietteCapel) [inline python_examples/intervention_example/social_distancing_simulation_flow.py]: What is this 17?

File: pyEpiabm/pyEpiabm/property/household_foi.py (hunk 1 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/property/household_foi.py]: I'm not sure if adding the 'distancing' in the foi actually influences the infectiousness. The result should be the same no matter this multiplication occurs in 'household_susc' or 'household_foi'.

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py (hunk 2 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py:27]: Done! I also done the similar thing in case isolation and place closure tests.

File: python_examples/intervention_example/social_distancing_simulation_flow.py (hunk 3 above)
Comment (jiayuanz3) [reply python_examples/intervention_example/social_distancing_simulation_flow.py]: There are 17 age groups, so the input should actually be [0, 0, 0, ... ]. Here I use a lazy way to implement, maybe better to replace 17 with len(pe.Parameters.instance().intervention_params['social_distancing'][to_modify_parameter])? Will it be too complex?

**Comment by HenrietteCapel** [comment]
We still don't have everything the same as CovidSim. We could add:
- Durations and delays can alternatively be specified at admin unit level.
- A trigger is defined flexibly, with reference to either the numbers (cumulative incidence) of confirmed cases, numbers of critical cases that require Intensive Care Units (ICU beds), or numbers of deaths. Numbers can be either absolute or per-capita, and the model may be refined so that the number is defined in relation to ICU bed capacity. The window over which cumulative incidence is calculated is specified by P.TriggersSamplingInterval [Number of sampling intervals over which cumulative incidence measured for global trigger] days. Further, triggers may be specified at either admin unit (regional) or at country (global) level. Currently, all interventions must use the same epidemiological variable for their trigger (e.g. all using confirmed cases or all using ICU beds), and interventions do not differ in whether they consider absolute or per-capita triggers.

File: python_examples/intervention_example/social_distancing_simulation_flow.py (hunk 3 above)
Comment (jiayuanz3) [reply python_examples/intervention_example/social_distancing_simulation_flow.py]: Changed numerical value 17 to num_age_group :)

File: pyEpiabm/pyEpiabm/core/microcell.py [hunk 4]
```diff
@@ -41,6 +41,7 @@ def __init__(self, cell):
         self.compartment_counter = _CompartmentCounter(
             f"Microcell {id(self)}")
         self.closure_start_time = None
+        self.distancing_start_time = None
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/core/microcell.py]: I wonder how many of these attributes you think there will be - we already have two related to interventions and I am not sure how many more we will need. If its a lot, maybe this isn't the best way to set this up? 

For example, could we consider not specifying these in the init file, and creating them on the fly if required? It would reduce the memory requirement for example

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+#
+# Social distancing Class
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/intervention/social_distancing.py]:
```suggestion
# Social Distancing Intervention
```

(fixing capitalisation, and because we know this is a class)

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+#
+
+import random
+
+from pyEpiabm.core import Parameters
+from pyEpiabm.intervention import AbstractIntervention
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/intervention/social_distancing.py]:
```suggestion
from pyEpiabm.core import Parameters

from .abstract_intervention import AbstractIntervention
```

check the local import works, but I think we settled on this style for imports within the same submodule. Its probably not an issue but can help you avoid circular import errors later down the line

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    ):
+        self.distancing_duration = distancing_duration
+        self.distancing_delay = distancing_delay
+        self.case_microcell_threshold = case_microcell_threshold
+        self.distancing_compliant = distancing_compliant
+        self.distancing_enhanced_prob = distancing_enhanced_prob
+        super(SocialDistancing, self).__init__(population=population,
+                                               **kwargs)
+
+    def __call__(self, time):
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/intervention/social_distancing.py:39]: I really don't understand what's going on with the age dependence here - if you have explained in the wiki, could you put a link in the docstring for this function (which you should also write!)

File: pyEpiabm/pyEpiabm/property/household_foi.py (hunk 1 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/property/household_foi.py]: True right now, because we only use the susceptibility in the foi. But if we end up using it for something else, and this should be in the susceptibility, then it would be good to put in there now

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        params = pe.Parameters.instance().intervention_params[
+            'social_distancing']
+        cls.socialdistancing = SocialDistancing(population=cls._population,
+                                                **params)
+
+    def test__init__(self):
+        self.assertEqual(self.socialdistancing.start_time, 6)
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py]: Are these valules default in the function, or from the params file? Would be good to document what you're actually testing here!

**Review by KCGallagher — CHANGES_REQUESTED** [review]
Broadly looks good! Have now read through all of it, and left some comments. I think you'll need a convincing argument to add so many lines of outputs files to the repo - maybe we just add one plot and the workflow, so the user can regenerate their own output files if they wish?

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py (hunk 2 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py:27]: Great suggestion @HenrietteCapel !

File: pyEpiabm/pyEpiabm/property/household_foi.py (hunk 1 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/property/household_foi.py]: I've moved distancing scaling step from foi function to susc function in household, place and spatial

File: pyEpiabm/pyEpiabm/core/microcell.py (hunk 4 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/core/microcell.py]: There are four start_time (each intervention has one). I think we have to initialise it rather than creating it on the fly because we are checking for example if 'microcell.distancing_start_time is not None' each time (e.g. in social_distancing.py). If we are not initialising it from the beginning, it contributes to attribute error.

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py (hunk 7 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/intervention/social_distancing.py:39]: I wrote a brief explanation in the docstring and also the put the wiki page link there (wiki hasn't finished yet).

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py (hunk 8 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py]: They are from params file, I added a comment line above to explain.

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import random
+
+from pyEpiabm.core import Parameters
+
+from .abstract_intervention import AbstractIntervention
+
+
+class SocialDistancing(AbstractIntervention):
+    """Social distancing intervention
+    Social distancing is based on the number of infecious persons in their
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/intervention/social_distancing.py]:
```suggestion
    Social distancing is based on the number of infectious persons in their
```

Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/intervention/social_distancing.py]: Do you also mean "each microcell" rather then "their microcells" - I'm not sure who is the subject here but finding this hard to parse!

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+                                               **kwargs)
+
+    def __call__(self, time):
+        for cell in self._population.cells:
+            for microcell in cell.microcells:
+                if microcell.distancing_start_time is not None:
+                    if time > microcell.distancing_start_time + self.\
+                              distancing_duration:
+                        # Stop social distancing after their distancing period
+                        microcell.distancing_start_time = None
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/intervention/social_distancing.py:47]: Does this disable distancing on the individual level too? I was expecting that to be handled here (so individuals are no longer distancing), unless that is only a microcell property (the docstring implies it is an individual property)

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                        for person in microcell.persons:
+                            if Parameters.instance().use_ages:
+                                r_age = random.random()
+                                if r_age < self.distancing_enhanced_prob[
+                                            person.age_group]:
+                                    person.distancing_enhanced = True
+                                else:
+                                    person.distancing_enhanced = False
+                            else:
+                                person.distancing_enhanced = False
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/intervention/social_distancing.py:62]: I can't see anything in the documentation that says enhanced distancing requires age-dependence, we should talk about this. I think it would be good to explain what enhanced distancing actually is (is it shielding in the context of the UK?) as its not currently clear why we have implemented two types of distancing

File: pyEpiabm/pyEpiabm/property/place_foi.py [hunk 12]
```diff
@@ -74,7 +74,22 @@ def place_susc(place, infector, infectee,
...
+                             'social_distancing'][
+                             'distancing_place_enhanced_susc'][place_idx]
+            else:
+                distancing = Parameters.instance().\
+                             intervention_params[
+                             'social_distancing'][
+                             'distancing_place_susc'][place_idx]
+        else:
+            distancing = 1
+        return place_susc * distancing
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/property/place_foi.py]:
```suggestion
        if infector.microcell.distancing_start_time is not None:
            if infector.distancing_enhanced is True:
                place_susc *= Parameters.instance().\
                              intervention_params[
                              'social_distancing'][
                              'distancing_place_enhanced_susc'][place_idx]
            else:
                place_susc *= Parameters.instance().\
                              intervention_params[
                              'social_distancing'][
                              'distancing_place_susc'][place_idx]
        return place_susc
```

This is slightly tidier I think (in general good to avoid cluttering the namespace if possible)

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py (hunk 8 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py]: Would it be possible to read these values in from the parameters file then, instead of hardcoding them? Would just mean that the tests won't fail if someone changes the parameters file down the line. You could mock the parameters if you prefered

**Review by KCGallagher — CHANGES_REQUESTED** [review]
Looks good - see also prev comments about output files on the other fork

File: pyEpiabm/pyEpiabm/core/microcell.py (hunk 4 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/core/microcell.py]: That makes sense, but its very easy to check if an attribute exists (`hasattr()`) which could be added to the if logic.

**Review by KCGallagher — COMMENTED** [review]
(see comments on SABS_R3 fork as well)

File: pyEpiabm/pyEpiabm/core/microcell.py (hunk 4 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/core/microcell.py]: Yeah, I removed initialisation for four interventions in person.py and microcell.py. Now I check if the start_time exists with hasattr() before checking if its value is None.

File: pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py (hunk 8 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_intervention/test_social_distancing.py]: Fixed!

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py (hunk 10 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/intervention/social_distancing.py:47]: As the distancing applies to all individuals in the microcell, I assign it as the microcell property (otherwise, the start time and duration are the same for each individual in the microcell). I think it will be redundant if write in the individual level.

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py (hunk 9 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/intervention/social_distancing.py]: I rewrote the docstring.

File: pyEpiabm/pyEpiabm/intervention/social_distancing.py (hunk 11 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/intervention/social_distancing.py:62]: I follow the structure in CovidSim (https://github.com/mrc-ide/covid-sim/blob/master/docs/intervention-description.md). It's kind of individual compliant to follow enhanced social distancing depending on their age group (I directly assume it as the probability). We can discuss more about it to see if it's necessary to divide the age group!

File: pyEpiabm/pyEpiabm/property/place_foi.py (hunk 12 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/property/place_foi.py]: Done

File: pyEpiabm/pyEpiabm/core/microcell.py (hunk 4 above)
Comment (HenrietteCapel) [reply pyEpiabm/pyEpiabm/core/microcell.py]: Okay! We indeed still need  to check if it is None because it is set to None after the policy is inactivated. Do the tests still run? As we are checking often if the value is None. Or did you change that as well?

File: pyEpiabm/pyEpiabm/property/spatial_foi.py [hunk 13]
```diff
@@ -96,11 +98,25 @@ def spatial_susc(susc_cell, infector, infectee, time: float):
...
+            if infector.distancing_enhanced is True:
+                distancing_susc = Parameters.instance().\
+                    intervention_params['social_distancing'][
+                        'distancing_spatial_enhanced_susc']
+            else:
+                distancing_susc = Parameters.instance().\
+                    intervention_params['social_distancing'][
+                        'distancing_spatial_susc']
+        else:
+            distancing_susc = 1
```
Comment (HenrietteCapel) [inline pyEpiabm/pyEpiabm/property/spatial_foi.py]: Would it be more clear to have distancing_susc = 1 above and than multiply if needed in the if loop. So similar as for spatial_susc?

File: pyEpiabm/pyEpiabm/tests/test_unit/test_property/test_place_foi.py [hunk 14]
```diff
@@ -67,15 +67,19 @@ def test_place_case_isolation(self):
...
+        # Place closure
         self.infector.microcell.closure_start_time = 1
         result_closure = PlaceInfection.place_inf(self.place, self.infector,
                                                   self.time)
         self.assertEqual(result_closure, 0)
 
     def test_place_household_quarantine(self):
-        # Not in quarantine (quarantine_start_time = None)
+        # Update place type, not in quarantine (quarantine_start_time = None)
+        self.infector.place_types.append(PlaceType.PrimarySchool)
```
Comment (HenrietteCapel) [inline pyEpiabm/pyEpiabm/tests/test_unit/test_property/test_place_foi.py]: What does this has to do with quarantine? Is this to make sure that the effect of place closure will not be applied to people that are in quarantine? If so than we should also check if they are not in isolation.

**Review by HenrietteCapel — COMMENTED** [review]
Looks good! See two small comments.

File: pyEpiabm/pyEpiabm/tests/test_unit/test_property/test_place_foi.py (hunk 14 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_property/test_place_foi.py]: Sorry, I made it the wrong place. It should under place_closure. Thanks!

File: pyEpiabm/pyEpiabm/property/spatial_foi.py (hunk 13 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/property/spatial_foi.py]: Good point! I also modified in household_foi.py.

File: pyEpiabm/pyEpiabm/core/person.py [hunk 15]
```diff
@@ -167,3 +165,11 @@ def remove_place(self, place):
             ind = place_list.index(place)
             self.places.pop(ind)
             self.place_types.remove(place.place_type)
+
+    def close_place(self, closure_place_type):
+        if (hasattr(self.microcell, 'closure_start_time') is True) and (
```
Comment (rccreswell) [inline pyEpiabm/pyEpiabm/core/person.py]:
```suggestion
        if (hasattr(self.microcell, 'closure_start_time')) and (
```
hasattr should be boolean already, no need to compare it to true

File: pyEpiabm/pyEpiabm/core/person.py [hunk 16]
```diff
@@ -167,3 +165,11 @@ def remove_place(self, place):
             ind = place_list.index(place)
             self.places.pop(ind)
             self.place_types.remove(place.place_type)
+
+    def close_place(self, closure_place_type):
```
Comment (rccreswell) [inline pyEpiabm/pyEpiabm/core/person.py:169]: Needs a docstring

File: pyEpiabm/pyEpiabm/intervention/place_closure.py [hunk 17]
```diff
@@ -7,39 +7,36 @@
 
 class PlaceClosure(AbstractIntervention):
     """Place closure intervention
-    Close places based on the number of icu patients and infecious persons
-    in their microcells and reopen places after their closure period.
+    Close specific types of places based on the number of infecious persons
```
Comment (rccreswell) [inline pyEpiabm/pyEpiabm/intervention/place_closure.py]:
```suggestion
    Close specific types of places based on the number of infectious persons
```

**Review by rccreswell — CHANGES_REQUESTED** [review]

File: pyEpiabm/pyEpiabm/intervention/place_closure.py (hunk 17 above)
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/intervention/place_closure.py]:
```suggestion
    Close specific types of places based on the number of infectious persons
```

File: pyEpiabm/pyEpiabm/core/microcell.py (hunk 4 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/core/microcell.py]: I've changed for all four interventions and conflicts will be solved when merging!

File: pyEpiabm/pyEpiabm/core/person.py (hunk 15 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/core/person.py]: Fixed it in all files!

File: pyEpiabm/pyEpiabm/core/person.py (hunk 16 above)
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/core/person.py:169]: Done!

**Review by rccreswell — APPROVED** [review]

File: pyEpiabm/pyEpiabm/tests/test_unit/test_property/test_place_foi.py (hunk 14 above)
Comment (HenrietteCapel) [reply pyEpiabm/pyEpiabm/tests/test_unit/test_property/test_place_foi.py]: Quarantine and isolation does not remove places from a person place_list, it only scales the foi. This means that if the place is also closed the foi is scaled twice. Does this make sense? Or should we check in place closure of people are not in quarantine/isolation? On the other hand if they are in quarantine/isolation there is a small chance that they still go and infect people and this chance should decrease further when place is closed as well.
